# 프로그래밍 맛보기

## 연습문제

**참고**

먼저 [프로그래밍 맛보기](https://codingalzi.github.io/code-workout-python/starting/) 내용을 공부한 다음에 연습문제를
풀어볼 것을 권장한다.

### 문제 1

먼저 아래와 같이 두 개의 변수를 선언한다.

In [1]:
width = 17
height = 12.0

두 변수를 이용하여 아래에 지정된 값을 가리키는 표현식을 작성한 다음에 실행 결과를 확인하라.

(1) width를 3으로 나눈 나머지

답:

In [2]:
width % 3

2

(2) width 나누기 3.0

답:

In [3]:
width / 3.0

5.666666666666667

(3) height 빼기 3

답:

In [4]:
height - 3

9.0

(4) height 곱하기 5에 7 더하기

답:

In [5]:
height * 5 + 7

67.0

(5) height 곱하기 5에 7.0 더하기

답:

In [6]:
height * 5 + 7.0

67.0

### 문제 2

아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정하라.
오류가 발생하지 않으면 실행 결과를 설명한다.

(1) 오류 수정 또는 실행 결과 설명

In [7]:
print("Hello World)

SyntaxError: unterminated string literal (detected at line 1) (2771583741.py, line 1)

답:

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 발생 원인: `print()` 함수의 인자로 사용된 문자열을 감싸는 큰따옴표가 시작 부분에만 사용되었음.

문자열을 닫는 큰따옴표를 추가하면 오류가 발생되지 않는다.

In [8]:
print("Hello World")

Hello World


(2) 오류 수정 또는 실행 결과 설명

In [9]:
print(21 8)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (1287684157.py, line 1)

답:

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 원인: `print()` 함수의 인자가 이상함. 
    예를 들어, `21 8` 또는 `21+8`을 잘못 적었을 가능성 있음.
    21과 8 사이에 공백이 있으면 어떤 의미인지 알 수 없음.

예을 들어 21+8을 잘못 적은 거라면 아래처럼 수정하면 해결된다.

In [10]:
print(21+8)

29


(3) 오류 수정 또는 실행 결과 설명

In [11]:
print(2++2)

4


답:

오류가 발생하지 않는다. 
이유는 `+2`가 `2`로 간주되어 결국 `2 + 2`가 계산되어 4가 화면에 출력된다.

### 문제 3

아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정하라.
오류가 발생하지 않으면 실행 결과를 설명한다.

(1) 오류 수정 또는 실행 결과 설명

In [12]:
x = 2;

답:

오류가 발생하지 않는다.
이유는 세미 콜론 `;` 기호가 파이썬에서는 굳이 필요 없지만 연속된 두 개의 명령문을 한 줄에 작성할 때 활용되기도 한다.
예를 들어 아래 코드에서처럼 다른 변수 할당 명령문을 이어서 작성할 수 있다.

In [13]:
x = 2; y = 5
print('x:', x)
print('y:', y)

x: 2
y: 5


세미 콜론 뒤어 다른 명령문이 오지 않으면 무시된다.
따라서 위 코드는 아래 코드와 동일한 일을 수행한다.

In [14]:
x = 2
print('x:', x)

x: 2


(2) 오류 수정 또는 실행 결과 설명

In [15]:
y = 3
z = x y

SyntaxError: invalid syntax (457711640.py, line 2)

답

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 원인: `x y`는 의미가 없는 표현식임. 곱셈, 나눗셈 등 연산자가 생략되었을 가능성이 큼.

In [16]:
y = 3
z = x * y

In [17]:
print(z)

6


(3) 오류 수정 또는 실행 결과 설명

In [18]:
even = 3 # 홀수 값
odd = 17 # 짝수 값
print(even + oddd)

NameError: name 'oddd' is not defined

답:

- 오류 종류: `NameError`. 이름 오류.
- 오류 발생 원인: 변수 `oddd`가 정의되어 있지 않음. 앞서 정의된 `odd`를 잘못 타이핑한 것으로 보임.

이미 선언된 `odd`를 `oddd` 대신 사용하면 오류가 발생하지 않는다.

In [19]:
even = 3 # 홀수 값
odd = 17 # 짝수 값
print(even + odd)

20


(4) 오류 수정 또는 실행 결과 설명

In [20]:
x = 3
zero = 0

print(x/zero)

ZeroDivisionError: division by zero

답:

- 오류 종류: `ZeroDivisionError`. 0으로 나누기 오류.
- 오류 원인: `zero` 변수가 0을 가리키기에 나눗셈의 분모가 0이되어 오류가 발생함.

아래 코드에서처럼 분모가 0이 되지 않도록 하면 오류가 해결된다.

In [21]:
x = 3
zero = 0

print(x/(zero + 1))

3.0


### 문제 4

(1) 아래 코드의 실행 결과를 설명하라.

In [22]:
5.2 * 7

36.4

답:

부동소수점과 소수점의 연산은 부동소수점으로 계산된다.

(2) 아래 표현식을 실행하면 오류가 발생하는 이유를 설명하라.

In [23]:
2.0 * "7.18"

TypeError: can't multiply sequence by non-int of type 'float'

답:

- 오류 종류: `TypeError`. 자료형 오류.
- 오류 발생 원인: 부동소수점과 문자열의 곱셈은 허용되지 않음.

부동소수점 2.0와 정수 2가 일반적으로는 그 기능이 다름을 잘 보여준다.
예를 들어 2.0 대신 2를 사용하면 오류가 발생하지 않는다

In [24]:
2 * "7.18"

'7.187.18'

연산자와 연산에 사용되는 값의 자료형에 따라 연산의 허용 여부가 결정된다는 점을 기억해 두어야 한다.

(3) 아래 표현식을 실행했을 때의 결과를 설명하라.

In [25]:
'3.14' + '2.71'

'3.142.71'

답:

문자열 두 개를 덧셈 기호로 연결한 표현식은 주어진 두 표현식을 이어붙힌 새로운 문자열을 표현한다.
따라서 위 코드는 `3.14`, `2.71` 두 문자열을 단순하게 이어붙인 `3.142.71` 문자열로 계산된다.

### 문제 5

오전 6시 52분에 집을 떠나서,
초반에는 km당 8분 15초의 속력으로 1km를 뛰고, 
이어서 km당 7분 12초의 속력으로 3km를 뛴 후, 
초반 속력으로 1km를 마저 달려 집으로 돌아왔을 때의 도착시간을 계산하는 코드를 작성하라.
단 집에 도착하는 시간을 아래와같이 출력되도록 해야 한다.

```
집에 도착 시간: 7 시 30 분 6 초
```

힌트: 시, 분, 초를 동시에 다루기 위해 계산을 초 단위로 통일시킨 후 아래 변수들을 선언하고 최종적으로 도착 시간을 지정된 형식으로 출력한다.

- `start`: 자정 기준 출발 시간까지 지난 시간
- `running_time`: 달린 시간
- `arrival_time`: 자정 기준 도착 시간까지 지난 시간

답:

(1) 초단위 도착시간 계산

시, 분, 초를 별도로 다루면 계산이 복잡해지기에 모든 값을 초 단위로 통일시킨다.
먼저 출발시간인 6시 52분이 밤 0시를 기준으로 몇 초 지난 시간인지
계산해서 `start` 변수에 할당한다.

In [26]:
start = (6 * 60 + 52) * 60
print(start)

24720


아래 두 변수는 1 킬로미터를 가볍게 달릴 때와 좀 빠르게 달릴 때 걸리는 시간(초)을 가리킨다.

In [27]:
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

그러면 달리는 시간을 초로 계산하면 2286초로 나온다.

In [28]:
running_time = slow_running + 3 * fast_running + slow_running
running_time

2286

도착시간은 출발시간에 달린 시간을 더하면 된다.

In [29]:
arrival = start + running_time
arrival

27006

(2) 시, 분, 초로 변환

이제 도착시간을 시, 분, 초로 변환하기 위해 1분은 60초, 1시간은 60분이라는 정보를
이용하여 도착시간에 대해 60으로 나눴을 때의 몫과 나머지 계산을 두 번 연속 적용한다.

- 도착시간의 초 계산: 60으로 나눴을 때의 나머지

In [30]:
arrival_seconds = arrival % 60
arrival_seconds

6

- 도착시간의 분 계산: 60으로 나눴을 때의 몫

In [31]:
arrival_minutes = arrival // 60
arrival_minutes

450

450분은 60분 보다 크기에 시와 분으로 다시 변환해야 한다.
한 시간이 60분이기에 역시 60으로 나눈 몫과 나머지를 계산한다.

- 도착시간의 시 계산: 60으로 나눴을 때의 몫

In [32]:
arrival_hours = arrival_minutes // 60
arrival_hours

7

- 도착시간의 분 계산: 60으로 나눴을 때의 나머지

In [33]:
arrival_minutes = arrival_minutes % 60
arrival_minutes

30

따라서 도착시간은 7시 30분 6초가 된다. 

In [34]:
print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


(3) 종합 정리

지금까지 설명한 모든 코드를 정리하면 다음과 같다.

In [35]:
start = (6 * 60 + 52) * 60
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

running_time = (1 + 1) * slow_running + 3 * fast_running
arrival = start + running_time

arrival_seconds = arrival % 60
arrival_minutes = arrival // 60
arrival_hours = arrival_minutes // 60
arrival_minutes = arrival_minutes % 60

print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초
